From 79496c80bf7797cee568330bdec9240fb1a273ee Mon Sep 17 00:00:00 2001
From: Alan Tull <atull@altera.com>
Date: Tue, 15 Jan 2013 10:03:37 -0600
Subject: [PATCH 185/254] arm: socfpga: enable designware i2c adapter

Upstream: git://git.rocketboards.org/linux-socfpga.git

 * Enable Designware I2C Adapter driver in defconfig
 * Add I2C adapter to device tree
 * Add I2C clks
 * By default, I2C adapters will default to fast mode.

Signed-off-by: Alan Tull <atull@altera.com>
(cherry picked from commit d09cda72893c1f3a9c942afc7c7d401e5e6eb971)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 arch/arm/boot/dts/socfpga.dtsi     |   18 ++++++++++++++++++
 arch/arm/configs/socfpga_defconfig |    4 ++++
 drivers/clk/socfpga/clk.c          |    8 ++++++++
 3 files changed, 30 insertions(+), 0 deletions(-)

diff --git a/arch/arm/boot/dts/socfpga.dtsi b/arch/arm/boot/dts/socfpga.dtsi
index e30e145..ed0759e 100644
--- a/arch/arm/boot/dts/socfpga.dtsi
+++ b/arch/arm/boot/dts/socfpga.dtsi
@@ -311,5 +311,23 @@
 				dev-tx-fifo-size = <512 512 512 512 512 512
 					512 512 512 512 512 512 512 512 512>;
 		};
+
+		i2c0: i2c@ffc04000 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			compatible = "snps,designware-i2c";
+			reg = <0xffc04000 0x1000>;
+			interrupts = <0 158 4>;
+			emptyfifo_hold_master = <1>;
+		};
+
+		i2c1: i2c@ffc05000 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			compatible = "snps,designware-i2c";
+			reg = <0xffc05000 0x1000>;
+			interrupts = <0 159 4>;
+			emptyfifo_hold_master = <1>;
+		};
 	};
 };
diff --git a/arch/arm/configs/socfpga_defconfig b/arch/arm/configs/socfpga_defconfig
index b3ee96b..3dcd5e9 100644
--- a/arch/arm/configs/socfpga_defconfig
+++ b/arch/arm/configs/socfpga_defconfig
@@ -109,3 +109,7 @@ CONFIG_DEBUG_INFO=y
 CONFIG_ENABLE_DEFAULT_TRACERS=y
 CONFIG_DEBUG_USER=y
 CONFIG_XZ_DEC=y
+CONFIG_I2C=y
+CONFIG_I2C_DESIGNWARE_CORE=y
+CONFIG_I2C_DESIGNWARE_PLATFORM=y
+CONFIG_I2C_CHARDEV=y
diff --git a/drivers/clk/socfpga/clk.c b/drivers/clk/socfpga/clk.c
index f7f8b1a..8d89542 100644
--- a/drivers/clk/socfpga/clk.c
+++ b/drivers/clk/socfpga/clk.c
@@ -102,6 +102,14 @@ void __init socfpga_init_clocks(void)
 			SOCFPGA_S2F_USR_CLK);
 	clk_register_clkdev(clk, "s2f_usr_clk", NULL);
 
+	clk = clk_register_fixed_rate(NULL, "i2c0_clk", NULL, CLK_IS_ROOT,
+			SOCFPGA_PER_PLL_CLK);
+	clk_register_clkdev(clk, NULL, "ffc04000.i2c");
+
+	clk = clk_register_fixed_rate(NULL, "i2c1_clk", NULL, CLK_IS_ROOT,
+			SOCFPGA_PER_PLL_CLK);
+	clk_register_clkdev(clk, NULL, "ffc05000.i2c");
+
 	clk = clk_register_gate(NULL, "gmac0_clk", "per_pll_clk", 0,
 			clk_mgr_base_addr + CLKMGR_PERPLLGRP_EN,
 			CLKMGR_EMAC0_CLK_EN, 0, &_lock);
-- 
1.7.5.4

